### PR TITLE
Audio: Optimize and fix TDFB direction calculation

### PR DESCRIPTION
**Pull request by ShriramShastry**
This check-in refines the TDFB direction calculation, addressing both
performance and correctness:

- **Fixes:**
- Correct infinite loop in `max_mic_distance` by fixing loop conditions.
- **Optimizations:**
- Fix off-by-one error in `line_array_mode_check` ensuring all checks
  for co-linearity among microphone locations are performed.

### CODE REVIEW

**Comment by singalsu** [comment]
It's not clear if you want to keep #9160, or want to include the previous change into this PR (and close #9160). I think separate PR is better, so you should keep #9160 and remove the first commit from this PR.

File: src/audio/tdfb/tdfb_direction.c
Comment (singalsu) [inline src/audio/tdfb/tdfb_direction.c:125]: Remove from commit text "Optimizations:
- Enhance `theoretical_time_differences` efficiency by substituting
  division with multiplication using a precomputed reciprocal.". That patch is elsewhere.

File: src/audio/tdfb/tdfb_direction.c
Comment (singalsu) [inline src/audio/tdfb/tdfb_direction.c:162]: Put this into a separate commit. Since this would be a bug fix. Previous change is an optimization for initialize code.

File: src/audio/tdfb/tdfb_direction.c
Comment (singalsu) [inline src/audio/tdfb/tdfb_direction.c:284]: Good  catch, without the Q1.15 multiply shifts this indeed fits the int32_t type. Please put this also to to own commit.

**Review by singalsu — CHANGES_REQUESTED** [review]

**Review by lgirdwood — COMMENTED** [review]
@ShriramShastry LGTM, but needs @singalsu to approve. Btw, what is the MCPS or % speed up ?

**Review by lyakh — CHANGES_REQUESTED** [review]

**Review by lgirdwood — COMMENTED** [review]
@ShriramShastry I think its best to split PR as suggested, to help with bisection. Some of the changes are ready to merge but are blocked by opens from other changes in the PR.

File: src/audio/tdfb/tdfb_direction.c
Comment (ShriramShastry) [reply src/audio/tdfb/tdfb_direction.c]: Done ! Please take a look. Thanks

File: src/audio/tdfb/tdfb_direction.c
Comment (lyakh) [inline src/audio/tdfb/tdfb_direction.c:130]: this breaks compilation and therefore bisection. No commit should break compilation.

**Review by lyakh — CHANGES_REQUESTED** [review]

File: src/audio/tdfb/tdfb_direction.c
Comment (lyakh) [inline src/audio/tdfb/tdfb_direction.c:128]: this comment doesn't help. Halves iterations - compared to what? The reader will not know what it was like before, they will only see the present state. So if you want any comment here you can just say, that you calculate distances between all possible microphone pairs.
In fact even in the commit description I don't think, that saying just "halves iterations" is enough, you need to explain why this works, why the other half isn't needed, i.e. that a distance between `i`th and `j`th microphones has to be calculated only once.

File: src/audio/tdfb/tdfb_direction.c
Comment (lyakh) [reply src/audio/tdfb/tdfb_direction.c:162]: btw, I think there's also a typo in the comment in line 158 - it should be `y(i)`, not `y(1)`

**Review by lyakh — COMMENTED** [review]
code changes look good now, just need to improve comments

File: src/audio/tdfb/tdfb_direction.c
Comment (lyakh) [inline src/audio/tdfb/tdfb_direction.c:130]: for the future: please make corrections, addressing review comments *in the commit where that change is made* - not in a later commit. It doesn't matter much in this case, since this is a comment, but it would make reviewing easier

**Review by lyakh — APPROVED** [review]

**Review by lgirdwood — COMMENTED** [review]
Lets wait for @singalsu to return before we merge.

File: src/audio/tdfb/tdfb_direction.c
Comment (marc-hb) [reply src/audio/tdfb/tdfb_direction.c:130]: Just FYI @lyakh: using new commits to address review comments is the "standard" GitHub process. Using Github like plain git is very unusual, SOF (and Zephyr) are among the outliers rewriting history and using "force" pushes.

https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/incorporating-feedback-in-your-pull-request#applying-suggested-changes

https://github.com/zephyrproject-rtos/zephyr/issues/39194

Just a warning that this issue will keep coming again and again.

Comment (lyakh) [reply src/audio/tdfb/tdfb_direction.c:130]: @marc-hb do I understand you correctly, that as soon as I push a PR with whatever bugs in my commits, the *standard* / recommended way to improve that PR would be to push fixes on top? So that after a merge you get my original 5 commits with all the bugs in them and then additional 3 commits on top with fixes? If so, I strongly disagree with this process. For one it would break bisection. And in general that isn't what a peer review process is about IMHO. Recall the "original LKML" (at least as of the git era) style review process - you send an email thread with your commits, you get a ton of replies explaining how wrong you are and suggesting all the therapists you should visit to address your problems, after which you go, fix your stuff and resend the emails - the original ones are gone, no fixing on top.
But importantly - we all make mistakes, very few PRs are recognised as good enough in their first version.

**Comment by singalsu** [comment]
> Lets wait for @singalsu to return before we merge.

We don't have CI validation for these features and @ShriramShastry is without DUTs to run tests. So bear with me, I need to test this manually before approve. Right now there is higher priority work for a new platform.

File: src/audio/tdfb/tdfb_direction.c
Comment (singalsu) [reply src/audio/tdfb/tdfb_direction.c:130]: Agree with @lyakh, This comment change should really be in the previous commit that just added it. It's not much work to shuffle changes from commit to other and @ShriramShastry should take the time to learn to clean up patches this way and ease our review work.

File: src/audio/tdfb/tdfb_direction.c
Comment (singalsu) [inline src/audio/tdfb/tdfb_direction.c:160]: Same, this part of code was changed in previous commit, would fit there better. The changes so far are initialize code changes (run once) and comments changes while the next one is good and actually reduces run-time MCPS.

**Review by singalsu — APPROVED** [review]
Approve due to bug fix  (would possibly treat a L-shape array as line array if last mic deviates from line) and a good optimization found.

Not so happy with commits structure with changes to previously added comments in next commit. But OK this time, please try to improve with those.

**Comment by singalsu** [comment]
Adding that this PR saves when sound angle tracking is enabled in TGL 4 mic case from 134 MCPS average to 120 MPCS average. Also test tools/test/audio/tdfb_direction_test.m is passed.

File: src/audio/tdfb/tdfb_direction.c
Comment (marc-hb) [reply src/audio/tdfb/tdfb_direction.c:130]: I think you missed the last sentence:

> Just a warning that this issue will keep coming again and again.

Comment (marc-hb) [reply src/audio/tdfb/tdfb_direction.c:130]: Just for fun, here's a very curious and much more extreme view: https://fossil-scm.org/home/doc/trunk/www/rebaseharm.md

**Comment by kv2019i** [comment]
Failures in https://sof-ci.01.org/sofpr/PR9169/build5921/devicetest/index.html , but the TDFB not part of these topologies, so can' be related. Proceeding with merge, thanks @ShriramShastry !

**Pull request merged**
